ReactiveCocoa/DelegateProxy.swift

```diff
@@ -1,106 +1,269 @@
+
+						let previousImpl = class_replaceMethod(perceivedClass, selector, _rac_objc_msgForward, types)
+						precondition(previousImpl == nil || previousImpl == _rac_objc_msgForward,
+						             "Swizzling DelegateProxy with other libraries is not supported. \(perceivedClass)")
+					}
+				}
 
-	required init(lifetime: Lifetime, _ originalSetter: @escaping (AnyObject) -> Void) {
-		self.lifetime = lifetime
-		self.originalSetter = originalSetter
+				objcProtocol.enumerateMethods(required: true, body: implement(required: true))
+				objcProtocol.enumerateMethods(required: false, body: implement(required: false))
+			}
+		}
 	}
 
-	override func forwardingTarget(for selector: Selector!) -> Any? {
-		return interceptedSelectors.contains(selector) ? nil : forwardee
+	open override func conforms(to aProtocol: Protocol) -> Bool {
+		return aProtocol === objcProtocol || super.conforms(to: aProtocol)
 	}
 
-	func intercept(_ selector: Selector) -> Signal<(), NoError> {
-		interceptedSelectors.insert(selector)
-		originalSetter(self)
-		return self.reactive.trigger(for: selector).take(during: lifetime)
+	open override func responds(to selector: Selector!) -> Bool {
+		return protocol_getMethodDescription(objcProtocol, selector, true, true).name != nil
+			|| interceptedSelectors.contains(selector)
+			|| NSObject.instancesRespond(to: selector)
+			|| forwardee?.responds(to: selector) ?? false
 	}
 
-	func intercept(_ selector: Selector) -> Signal<[Any?], NoError> {
+	internal final func proxyWillIntercept(_ selector: Selector) {
+		writeLock.lock()
+		defer { writeLock.unlock() }
 		interceptedSelectors.insert(selector)
 		originalSetter(self)
-		return self.reactive.signal(for: selector).take(during: lifetime)
 	}
+}
 
-	override func responds(to selector: Selector!) -> Bool {
-		if interceptedSelectors.contains(selector) {
-			return true
+extension Protocol {
+	fileprivate func enumerateMethods(required: Bool, body: (Selector, UnsafePointer<Int8>) -> Void) {
+		var count: UInt32 = 0
+		if let list = protocol_copyMethodDescriptionList(self, required, true, &count) {
+			UnsafeBufferPointer(start: list, count: Int(count))
+				.lazy.filter { $0.name != nil }.forEach { body($0.name, $0.types) }
+			free(list)
 		}
+	}
+}
 
-		return (forwardee?.responds(to: selector) ?? false) || super.responds(to: selector)
+internal func unsafeDelegateProxy(_ proxy: Unmanaged<NSObject>, didInvoke selector: Selector, with invocation: AnyObject) {
+	let proxy = proxy.takeUnretainedValue() as! _DelegateProxyProtocol
+	if let forwardee = proxy._forwardee, forwardee.responds(to: selector) {
+		invocation.invoke(withTarget: forwardee)
 	}
 }
 
-private let hasSwizzledKey = AssociationKey<Bool>(default: false)
+internal func isDelegateProxy(_ type: AnyClass) -> Bool {
+	return type is _DelegateProxyProtocol.Type
+}
+
+extension Reactive where Base: NSObject, Base: DelegateProxyProtocol {
```

Comment:
I think these methods could go on `DelegateProxy` directly.

Reply:
I agree that a shorthand on `DelegateProxy` would be nice. But what the overriding is about is that `reactive.trigger(for:)` is available on all `NSObject`, thus also being available on `DelegateProxy`.